packages/appium/lib/schema/schema.js

```diff
@@ -0,0 +1,571 @@
+   * @see https://npm.im/ajv-formats
+   * @private
+   */
+  constructor () {
+    this._ajv = AppiumSchema._instantiateAjv();
+  }
+
+  /**
+   * Factory function for {@link AppiumSchema} instances.
+   *
+   * Returns a singleton instance if one exists, otherwise creates a new one.
+   * Binds public methods to the instance.
+   * @returns {AppiumSchema}
+   */
+  static create () {
+    const instance = AppiumSchema._instance ?? new AppiumSchema();
+    AppiumSchema._instance = instance;
+
+    _.bindAll(instance, [
+      'finalize',
+      'flatten',
+      'getArgSpec',
+      'getDefaults',
+      'getSchema',
+      'hasArgSpec',
+      'isFinalized',
+      'registerSchema',
+      'reset',
+      'validate',
+    ]);
+
+    return instance;
+  }
+
+  /**
+   * Returns `true` if a schema has been registered using given extension type and name.
+   *
+   * This does not depend on whether or not the instance has been _finalized_.
+   * @param {ExtensionType} extType - Extension type
+   * @param {string} extName - Name
+   * @returns {boolean} If registered
+   */
+  hasRegisteredSchema (extType, extName) {
+    return this._registeredSchemas[extType].has(extName);
+  }
+
+  /**
+   * Return `true` if {@link AppiumSchema.finalize finalize} has been called
+   * successfully and {@link AppiumSchema.reset reset} has not been called since.
+   * @returns {boolean} If finalized
+   */
+  isFinalized () {
+    return Boolean(this._finalizedSchemas);
+  }
+
+  /**
+   * Call this when no more schemas will be registered.
+   *
+   * This does three things:
+   * 1. It combines all schemas from extensions into the Appium config schema,
+   *    then adds the result to the `Ajv` instance.
+   * 2. It adds schemas for _each_ argument/property for validation purposes.
+   *    The CLI uses these schemas to validate specific arguments.
+   * 3. The schemas are validated against JSON schema draft-07 (which is the
+   *    only one supported at this time)
+   *
+   * Any method in this instance that needs to interact with the `Ajv` instance
+   * will throw if this method has not been called.
+   *
+   * If the instance has already been finalized, this is a no-op.
+   * @public
+   * @throws {Error} If the schema is not valid
+   * @returns {Readonly<Record<string,StrictSchemaObject>>} Record of schema IDs to full schema objects
+   */
+  finalize () {
+    if (this.isFinalized()) {
+      return /** @type {Record<string,StrictSchemaObject>} */ (
+        this._finalizedSchemas
+      );
+    }
+
+    const ajv = this._ajv;
+
+    // Ajv will _mutate_ the schema, so we need to clone it.
+    const baseSchema = _.cloneDeep(/** @type {StrictSchemaObject} */(appiumConfigSchema));
+
+    /**
+     *
+     * @param {SchemaObject} schema
+     * @param {ExtensionType} [extType]
+     * @param {string} [extName]
+     */
+    const addArgSpecs = (schema, extType, extName) => {
+      _.forEach(schema, (propSchema, propName) => {
```

Comment:
as a rule we use for loops instead of `_.forEach`

Reply:
we should create an eslint rule about it 😄 

anyway, I used `_.forEach()` because we are iterating over an object, not an array.  I suppose I could use for..of `Object.entries` or something, assuming Node v12 supports it.

Reply:
this is changed